### PR TITLE
Multi cursor visual maps

### PR DESCRIPTION
**Pull request by djjcast**
Here's a fix for issue #111.

### CODE REVIEW

**Comment by djjcast** [comment]
Hmm, that's weird, all tests are passing on my system (MacVim). Are the tests being run in terminal Vim? Maybe it's failing because of a lag/synchronization issue in terminal Vim?

**Comment by balta2ar** [comment]
I've ran all tests locally both in gvim and vim (hacked vimrunner's gem a bit) and they all have passed.

**Comment by balta2ar** [comment]
@eapache Any ideas?

File: autoload/multiple_cursors.vim [hunk 1]
```diff
@@ -1083,8 +1086,8 @@ endfunction
 let s:retry_keys = ""
 function! s:display_error()
   if s:bad_input == s:cm.size()
-        \ && s:from_mode ==# 'n'
-        \ && has_key(g:multi_cursor_normal_maps, s:char[0])
+        \ && ((s:from_mode ==# 'n' && has_key(g:multi_cursor_normal_maps, s:char[0]))
+        \ ||  (s:from_mode ==# 'v' && has_key(g:multi_cursor_visual_maps, s:char[0])))
```
Comment (eapache) [inline autoload/multiple_cursors.vim]: Should this be checking for visual block mode as well? I think that's `V`, but I'm not sure if the same multi-key mappings should work in it?

**Comment by eapache** [comment]
One question, otherwise LGTM.

The test suite seems to be flaky for some reason - I retried it and a different test failed. Everything passed the third time around. Everything passes for me locally though, I wonder if it's something to do with the CI changes I made recently in #183 and #184

**Comment by djjcast** [comment]
Yes, the multi-key commands should work in linewise Visual mode. The check has been updated.

**Comment by balta2ar** [comment]
I've given your awesome fix a shot and I've discovered few conflicts. Specifically, there are conflicts with [targets](https://github.com/wellle/targets.vim) and [vim-fanfingtastic](https://github.com/dahu/vim-fanfingtastic) plugins.

### Minimal vimrc

Here is my minimal initial vimrc where everything is working fine:
``` vim
if has('vim_starting')
   set nocompatible
   set runtimepath+=~/.vim/bundle/neobundle.vim/
endif

call neobundle#begin(expand('~/.vim/bundle/'))
NeoBundleFetch 'Shougo/neobundle.vim'
NeoBundle 'wellle/targets.vim'
"NeoBundle 'dahu/vim-fanfingtastic'
NeoBundle 'terryma/vim-multiple-cursors'
call neobundle#end()
NeoBundleCheck

"let g:targets_aiAI = 'aIAi'
```
### Input file
Here is the input file that I use for all cases below:
```
hello (abc apple) world
```
### Command sequence
Combination of commands that I use to reproduce the problem: `fa<C-n>i(`.

### Case 1

Uncomment `NeoBundle 'dahu/vim-fanfingtastic'`. After applying command sequence, words inside parenthesis are selected **but** newline after the first line is inserted right after `i` is pressed:
```
hello (abc apple) world
<empty new line here>
```

### Case 2

Uncomment `let g:targets_aiAI = 'aIAi'`. This is the option for targets that I use to swap the meaning of `i` and `I` commands. You can keep fanfingtastic uncommented or not, looks like result is the same. After applying command sequence, garbage is inserted into the first line and parenthesis is inserted into the beginning of the second line:

```
(multiple-cursors-apply):call <SNR>32_detect_bad_input()
(hello (abc apple) world
```
### Summary
I'm not sure which plugin to blame, maybe both targets and fanfingtastic. @djjcast, could you look into this, please?

**Comment by djjcast** [comment]
@balta2ar vim-multiple-cursors is incompatible with plugins that map keys to multi-key sequences.

For example, targets.vim creates this mapping: `onoremap  i( :<C-U>call targets#o('(ci', v:count1)<CR>`

So typing `i(` triggers the input of `:<C-U>call targets#o('(ci', v:count1)<CR>` and vim-multiple-cursors applies this input one character at a time to each cursor, i.e., `:` is applied to each cursor, `<C-u>` is applied to each cursor, `c` is applied to each cursor, etc.

vim-multiple-cursors would have to know that when you type `i(`, you want `:<C-U>call targets#o('(ci', v:count1)<CR>` to be applied to each cursor as an atomic operation. So neither plugin is to blame - it's a compatibility issue that would probably be nontrivial to fix.

The MAPPINGS section of `:help write-plugin` explains how mappings are processed.

As a workaround you could temporarily remap `i(` like this:
```VimL
" Called once right before you start selecting multiple cursors
function! Multiple_cursors_before()
  onoremap <buffer> i( i(
endfunction

" Called once only when the multiple selection is canceled (default <Esc>)
function! Multiple_cursors_after()
  ounmap <buffer> i(
endfunction
```

**Comment by balta2ar** [comment]
@djjcast Thank you for the detailed explanation.

> it's a compatibility issue that would probably be nontrivial to fix.

Still, could you think (just think) of a solution for this issue? For example, can this plugin look ahead and see the whole input sequence `:<C-U>call targets#o('(ci', v:count1)<CR>`, scan it, detect, that there is `<C-U>`, and use that fact as an indicator that the whole sequence needs to be applied atomically?

**Comment by djjcast** [comment]
It seems like it's not possible to capture input from a mapping. Here's a test:
```VimL
function! CaptureInput()
    echomsg 'enter CaptureInput()'

    let char = ''
    while char != "\<Esc>"
        let char = nr2char(getchar())
        echomsg char
    endwhile

    echomsg 'exit CaptureInput()'
endfunction
nnoremap <Plug>(CaptureInput) :<C-u>call CaptureInput()<CR>

function! GetChar()
    echomsg 'enter GetChar()'

    " get the input character `e`
    let char = nr2char(getchar())
    echomsg 'character input: ' . char

    " capture any input after the mapping for `e` is invoked
    call feedkeys(char . "\<Plug>(CaptureInput)")

    echomsg 'exit GetChar()'
endfunction

nnoremap e :<C-u>echomsg 'hello, world'<CR>
```
Add the above code to a file and `:source` it. Now `:call GetChar()` and type `e<Esc>`. Unfortunately, `CaptureInput()` doesn't see the `:<C-u>echomsg 'hello, world'<CR>`.

I seems like after `feedkeys()` is called, Vim's input buffer is `["e", "\<Plug>(CaptureInput)"]`. Then when Vim realizes `e` is mapped, the input buffer becomes `[":<C-u>echomsg 'hello, world'<CR>", "\<Plug>(CaptureInput)"]` instead of `["\<Plug>(CaptureInput)", ":<C-u>echomsg 'hello, world'<CR>"]` so `CaptureInput()` doesn't see the mapping. Or maybe mappings are transformed internally instead of being add to the input buffer.

**Pull request merged**